### PR TITLE
ALAN Manual: Container Limits

### PR DESCRIPTION
**Pull request by tajmone**
@thoni56, I'm using this PR to track the changes proposed in #111 regarding `Container Limits` and the need to mention how nested containers are treated.

Before merging into `master`, we should discuss the details and keep adding changes to the `man-container-limits` branch, until we're happy the contents are fine, then squash all commits into a single one before merging.

I'm setting the PR as a "draft", and you as its reviewer, so that we don't accidentally merge it in unfinished state.passages.

### CODE REVIEW

File: manual/manual_03.asciidoc
Comment (tajmone) [inline manual/manual_03.asciidoc:1725]: @thoni56, this note is not very clear. I've added some comments above the admonition/NOTE to illustrate the unclear points.

**Review by tajmone — COMMENTED** [review]
# Limits BNF Rule Is Unclear

@thoni56, the BNF rule for `Limits` in the _Manual_ is somehow unclear:

```bnf
limits = 'LIMITS' {limit}

limit = limiting_attribute 'ELSE' {statement}

limiting_attribute = attribute_definition
                   | 'COUNT' integer
```

since the above `{limit}` implies that zero or more limits definitions can follow the `LIMITS` keyword — but the Manual doesn't mention what happens when there's just an empty `Limits` block, with no specifications. 

Does is clear inherited `Limits`? Is that what it should be used for?

In that case, it might be worth mentioning it.

This the actual BNF rule from [compiler/alan.g][alan.g]:

```bnf
optional_limits :
                | 'limits' limits
                   ;

limits : limit
       | limits limit
          ;

limit : limit_attribute else_or_then statements
         ;

else_or_then : 'else'
             | 'then'
                ;

limit_attribute : attribute_definition
                | 'count' INTEGER
                   ;
```

from the above rule it seems that although the whole construct is optional, a `Limits` keyword needs to be followed by at least one `limits` definition.

So I was wondering whether the BNF rules in the _Manual_ shouldn't actually be changed to:

```bnf
limits = 'LIMITS' limit {limit}

limit = limiting_attribute 'ELSE' {statement}

limiting_attribute = attribute_definition
                   | 'COUNT' integer
```

which would imply at least one `limit` definition.

Also, I noticed that the original BNG grammar allows also using `THEN` in place of `ELSE`, which is also not mentioned in the _Manual_ (neither in the BNF rules nor the text).  



[alan.g]: https://github.com/alan-if/alan/blob/master/compiler/alan.g#L345

File: manual/manual_03.asciidoc
Comment (thoni56) [inline manual/manual_03.asciidoc:1682]: I don't like the word "enforce" when it comes to what the author does. He only defines, so maybe that is a better word. The enforcing of those limits is done by the interpreter.

File: manual/manual_03.asciidoc
Comment (thoni56) [inline manual/manual_03.asciidoc:1700]: It is not "in the middle of the execution". It is exactly in that Locate statement, which might be anywhere. Maybe just remove the "in the middle of its execution" so it becomes "... be interrupted by a 'Locate' statement ..."

**Review by thoni56 — CHANGES_REQUESTED** [review]

**Comment by thoni56** [comment]
> Also, I noticed that the original BNG grammar allows also using `THEN` in place of `ELSE`, which is also not mentioned in the _Manual_ (neither in the BNF rules nor the text).

(From memory:) I think the initial grammar used `THEN` but as that is a positive word, used in `IF`, `WHEN` and `CASE` to indicate "do this now", it was exchanged for `ELSE` which indicates "negative" and goes more in hand with the "otherwise do this" of `IF`, `CASE`, `CHECK`  and restrictions. So `THEN` is just in the grammar for the compiler for backwards compatibility. (But should probably be removed now.)

**Review by thoni56 — APPROVED** [review]

**Comment by thoni56** [comment]
(I haven't done PR reviews in GitHub before...)

It felt kind of silly to review my own changes, and I couldn't figure out how to allow @tajmone to become a review of his own PR ;-) so...

I changed the three NOTES to normal paragraphs, it felt out of place to have three... I also added a WARNING about the interrupted execution after the section on Extract.

Could @tajmone have a look and perhaps concur that we can merge this now?

**Review by thoni56 — APPROVED** [review]

**Comment by tajmone** [comment]
> It felt kind of silly to review my own changes, and I couldn't figure out how to allow @tajmone to become a review of his own PR ;-) so...

I know, it's odd. I can't assign myself as a reviewer because I'm the creator of the PR, so the rule seems that PR creators can't be reviewers. Strangely though, I can dismiss any open review, probably due to repo privileges. I need to read the docs about PR reviews.

> Could @tajmone have a look and perhaps concur that we can merge this now?

I just wanted to polish the sentence "including any instances in containers that might be in the container that is added to", which I commented upon — but couldn't open a review for — so I'm not sure if you're seeing it, it's in the "File changes" tab. 

I'll try and fix it now, it's just that I wasn't sure what the sentence is actually trying to say, it's a bit entangled.

After that, I'll squash all commits locally, add a clean message and force push before merging (squash-merging from the WebUI could result in a strange commit message, I've read many complaints on how it works).

**Comment by thoni56** [comment]
> I just wanted to polish the sentence "including any instances in containers that might be in the container that is added to", which I commented upon — but couldn't open a review for — so I'm not sure if you're seeing it, it's in the "File changes" tab.
> 
> I'll try and fix it now, it's just that I wasn't sure what the sentence is actually trying to say, it's a bit entangled.

I don't actually see that comment. I'll try to rephrase now and let's see if you think it's an improvement ;-)

**Comment by tajmone** [comment]
# Seems Ready to Merge!

> I don't actually see that comment. I'll try to rephrase now and let's see if you think it's an improvement ;-)

Yes, that nails it! much clearer.

I've squashed all commits in this PR branch into a single commit with the original message, so if you're OK with I propose to rebase-merge the PR!

PS: Sorry for the delay in replying, but a friend came to visit from abroad so we spent the whole day together.

**Pull request merged**
